algorithms/HPdclassifier/R/hpdegbm.R

```diff
@@ -0,0 +1,290 @@
+         group = NULL)
+
+       # estimate the best iteration
+       best.iter0 <- gbm.perf(dGBM_model, 
+              plot.it = FALSE, 
+              oobag.curve = FALSE, 
+              overlay = FALSE, 
+              method="OOB")
+
+
+       # gbm.more and gbm.perf have bugs for multinomial distribution 
+       best.iter <- as.matrix(best.iter0)
+
+
+       # convert dGBM_model to a list
+       dGBM_modeli <- list(dGBM_model) 
+  
+       update(dGBM_modeli)
+       update(best.iter)
+      })
+    } else{  ### For big data: load each partition into every core
+     # X_train: dframe/darray
+     # Y_train: darray
+     # if nExecutor > npartition_train, distributed sampling can generate nExecutor partitions. nExecutor=npartition_of_sampled X_train. Or use one partition multiple times by i%%npartition_train+1
+     npartition_train <- npartitions(X_train)
+     # nExecutor <- npartition_train # already through random sampling
+     foreach(i, 1:nExecutor, function(dGBM_modeli=splits(dl_GBM_model,i), best.iter=splits(dbest.iter,i), x=splits(X_train,i%%npartition_train+1),y=splits(Y_train,i%%npartition_train+1),
+                  n.trees=n.trees, distribution=distribution, interaction.depth=interaction.depth, n.minobsinnode=n.minobsinnode, 
+                  shrinkage=shrinkage, bag.fraction=bag.fraction) {
+         library(gbm)
+
+         if (distribution=="multinomial") {
+            y <- unlist(y) #  convert it back to "factor" for multinomial distribution
+         }
+
+         ### apply gbm.fit for GBM modeling: local GBM model
+         dGBM_model <- gbm.fit(x, y,  
+            offset = NULL, 
+            misc = NULL, 
+            distribution = distribution,
+            w = NULL,
+            var.monotone = NULL,
+            n.trees = n.trees, 
+            interaction.depth = interaction.depth, 
+            n.minobsinnode = n.minobsinnode,
+            shrinkage = shrinkage,
+            bag.fraction = bag.fraction, 
+            nTrain = NULL,
+            train.fraction = NULL,
+            keep.data = TRUE,
+            verbose = TRUE,
+            var.names = NULL,
+            #response.name = "y",
+            group = NULL)
+
+          best.iter0 <- gbm.perf(dGBM_model, 
+              plot.it = FALSE, 
+              oobag.curve = FALSE, 
+              overlay = FALSE, 
+              method="OOB")
+
+       
+          best.iter <- as.matrix(best.iter0)
+
+          dGBM_modeli <- list(dGBM_model) 
+  
+          update(dGBM_modeli)
+          update(best.iter)
+        })
+    }
+  
+    if(trace) {
+	timing_info <- Sys.time() - timing_info
+	print(timing_info)
+    }
+
+    # output of hpdegbm: GBM model and best iteration estiamtion
+    GBM_model1 <- getpartition(dl_GBM_model)
+    best.iter1 <- getpartition(dbest.iter)
+
+    if (completeModel) {
+       finalModel <- list(GBM_model1,best.iter1, timing_info)
+    } else {
+       finalModel <- list(GBM_model1,best.iter1)
+    }
```

Comment:
We may keep completeModel interface argument. If we want to add something into the model for  future algorithms in the R gbm package, it will be more convenient to have completeModel.